docs/dev_guide/frequently-asked-questions-faq.rst

```diff
@@ -162,6 +163,98 @@ expected results. Instead, exclude CUB library implementation source code from
 your migration by adding ``--in-root-exclude=<path to CUB library source code>``
 to your migration command.
 
+How do I fix the hang issue of sycl code cased by work group level synchronization like group barrier used in conditional statement?
+************************************************************************************************************************************
+
+In SYCL code, if there is calling of synchronization API in control flow statements
+like conditional statement and loop statement, you may encounter runtime hang issue.
+The basic idea to fix the hang issue is to ensure that each synchronization API is
+either reached by all work items of a workgroup or skipped by all the work items of
+a workgroup.
+
+Here are two fix examples for reference:
+
+In the first example, synchronization API group barrier(nd_item.barrier()) is called
+inside an if block and the evaluation results of the conditional statement differs in
+each work item, not all work items can reach the synchronization API.
+
+.. code-block:: cpp
+   :linenos:
+
+   // original code
+
+   void kernel(const sycl::nd_item<3> &item_ct1) {
+      unsigned int tid = item_ct1.get_local_id(2);
+      if (tid < 32) {
+         ...
+         item_ct1.barrier(sycl::access::fence_space::local_space);
+         ...
+      }
+   }
+
+The following manually-adjusted code shows how to fix the hang issue
+or crash by moving the synchronization statement out of the if block.
+
+.. code-block:: cpp
+   :linenos:
+
+   // fixed SYCL code
+
+   void kernel(const sycl::nd_item<3> &item_ct1) {
+      unsigned int tid = item_ct1.get_local_id(2);
+      if (tid < 32) {
+         ...
+      }
+      item_ct1.barrier(sycl::access::fence_space::local_space);
+      if (tid < 32) {
+         ...
+      }
+   }
+
+The second example demonstrates how to fix the hang issue when
+a synchronization API is used in a for loop:
+
+.. code-block:: cpp
+   :linenos:
+
+   // original code
+
+   void compute(int id_space, const sycl::nd_item<3> &item_ct1) {
+      unsigned int id =
+         item_ct1.get_group(2) * item_ct1.get_local_range(2) +
+         item_ct1.get_local_id(2);
+      for (; id < id_space; id +=
+         item_ct1.get_group_range(2) * item_ct1.get_local_range(2)) {
+         ...
+         item_ct1.barrier();
+         ...
+      }
+   }
+
+The following is the manually adjusted code which ensures all work
+items has same run footprint in the for loop.
+
+.. code-block:: cpp
+   :linenos:
+
+   // fixed SYCL code
+
+   void compute(int id_space, const sycl::nd_item<3> &item_ct1) {
+      unsigned int id =
+         item_ct1.get_group(2) * item_ct1.get_local_range(2) +
+         item_ct1.get_local_id(2);
+      unsigned int num_workitem = 
+               item_ct1.get_group_range(2) * item_ct1.get_local_range(2);
+      // The condition is rounded up to an integer multiple of the num of work items

```

Comment:
the condition is updated to make sure all work items can enter the loop body in each iteration